### PR TITLE
pcapng enhancements (idb,epb) and some fixes

### PR DESCRIPTION
**Pull request by jcpvdm**
**Checklist:**

-   [x] If you are new to Scapy: I have checked [CONTRIBUTING.md](https://github.com/secdev/scapy/blob/master/CONTRIBUTING.md) (esp. section submitting-pull-requests)
-   [x] I squashed commits belonging together
-   [x] I added unit tests or explained why they are not relevant
-   [x] I executed the regression tests (using `cd test && ./run_tests` or `tox`)
-   [ ] If the PR is still not finished, please create a [Draft Pull Request](https://github.blog/2019-02-14-introducing-draft-pull-requests/)

Based on draft-ietf-opsawg-pcapng-latest, 5 March 2024, I made following changes:
-  map `packet.sniffed_on` to unique IDB id. When writing, if_name option and linktype is populated based on the first packet seen with a unique sniffed_on string.
-  map `packet.direction` to EPB flags inbound/outbound direction bits. Remaining flag bits not implemented.
-  simplified `RawPcapNgReader._read_options()` and moved (code,value) treatment to the caller since codes of same type can have different meaning for different type of blocks.
- Fix `RawPcapNgReader._read_block_shb()` and `_write_block_shb()` to be according to the draft
- Added new tests and adapted others to the changes.

Note: I couldn't figure the meaning of `packet.direction `on Scapy API reference or source code, but since it's marked as integer I assumed it follows the pcapng epb_flags outbound/inbound field.

Why I did this changes:
I was using scapy to convert packet hex dump from a text file to pcap. The text file included information about interface name and direction (outbound/inbound) which I would like to to have included in the pcapng file.

### CODE REVIEW

File: test/regression.uts [hunk 1]
```diff
@@ -2156,13 +2157,12 @@ pktcapng = sniff(offline=scapy_path("/test/pcaps/macos.pcapng.gz"))
 assert len(pktcapng) != 0
 
 = Write a pcapng
-
```
Comment (guedou) [inline test/regression.uts:2159]: Please refrain from adding un-necessary changes.

File: test/regression.uts [hunk 2]
```diff
@@ -2453,7 +2503,6 @@ l = [ p for p in RawPcapReader(f) ]
 assert len(l) == 1
 
 = Check RawPcapReader on pcap
-~ pcap
```
Comment (guedou) [inline test/regression.uts:2456]: Can you explain why you are removing multiple `~ pcap` lines?

File: scapy/utils.py [hunk 3]
```diff
@@ -1636,12 +1639,21 @@ def _read_block_shb(self):
             raise EOFError
 
         blocklen = struct.unpack(self.endian + "I", _blocklen)[0]
-        if blocklen < 16:
-            warning("Invalid SHB block length!")
+        if blocklen < 28:
```
Comment (guedou) [inline scapy/utils.py:1642]: Can you explain this change? Is it an error in the current implementation?

File: scapy/utils.py [hunk 4]
```diff
@@ -1636,12 +1639,21 @@ def _read_block_shb(self):
...
+            warning(f"Invalid SHB block length ({blocklen})!")
             raise EOFError
-        options = self.f.read(blocklen - 16)
-        self._read_block_tail(blocklen)
+        # major version must be 1
```
Comment (guedou) [inline scapy/utils.py:1645]:
```suggestion

        # Major version must be 1
```

File: scapy/utils.py [hunk 5]
```diff
@@ -1636,12 +1639,21 @@ def _read_block_shb(self):
...
+        _major = self.f.read(2)
+        major = struct.unpack(self.endian + "H", _major)[0]
+        if major != 1:
+            warning(f"SHB Major version {major} unsupported !")
+            raise EOFError
+        # skip minor version & section length
```
Comment (guedou) [inline scapy/utils.py:1651]:
```suggestion

        # Skip minor version & section length
```

File: scapy/utils.py [hunk 6]
```diff
@@ -1657,23 +1669,20 @@ def _read_packet(self, size=MTU):  # type: ignore
 
     def _read_options(self, options):
         # type: (bytes) -> Dict[str, Any]
-        """Section Header Block"""
-        opts = self.default_options.copy()  # type: Dict[str, Any]
+        opts = dict()
```
Comment (guedou) [inline scapy/utils.py:1672]: Please explain this change.

File: scapy/utils.py [hunk 7]
```diff
@@ -1657,23 +1669,20 @@ def _read_packet(self, size=MTU):  # type: ignore
...
-                opts["comment"] = options[4:4 + length]
+
+            if code != 0 and 4 + length < len(options):
+                opts[code] = options[4:4 + length]
+
             if code == 0:
                 if length != 0:
-                    warning("PcapNg: invalid option length %d for end-of-option" % length)  # noqa: E501
+                    warning("PcapNg: invalid option "
+                            "length %d for end-of-option" % length)  # noqa: E501
```
Comment (guedou) [inline scapy/utils.py:1685]: Is noqa still useful here?

File: test/regression.uts (hunk 1 above)
Comment (jcpvdm) [reply test/regression.uts:2159]: thanks for feedback! I will try avoid it in future.

File: test/regression.uts (hunk 2 above)
Comment (jcpvdm) [reply test/regression.uts:2456]: While testing, noticed that some pcap tests didn't have the `~ pcap` keyword, so I added the keyword manually for some of them. Then discovered that keyword could be simply applied to the whole test set instead, so added this:
```
############
############
+ pcap / pcapng format support
~ pcap
```
Finally when squashing the commits, I though it would be ok to simply delete the `~ pcap` keyword from individual tests, given that it was covered already by the test set.

File: scapy/utils.py (hunk 3 above)
Comment (jcpvdm) [reply scapy/utils.py:1642]: From [SHB specification ](https://ietf-opsawg-wg.github.io/draft-ietf-opsawg-pcap/draft-ietf-opsawg-pcapng.html#name-section-header-block ) i understand the mandatory fields add up to 28 bytes.

File: scapy/utils.py (hunk 6 above)
Comment (jcpvdm) [reply scapy/utils.py:1674]: _read_options() function is called by _read_block_shb(), _read_block_idb() and _read_block_epb()
The option codes may have different meaning for different blocks (eg: code 2 mean if_name for IDB, while for EPB it means epb_flags)
So, i thought it would be better to simplify this function to just read generic codes, and leave the treatment and default option behavior to the callers.

File: scapy/utils.py (hunk 7 above)
Comment (jcpvdm) [reply scapy/utils.py:1687]: I think it can be removed or the change undone.

File: scapy/utils.py [hunk 8]
```diff
@@ -1636,12 +1639,23 @@ def _read_block_shb(self):
...
+        if major != 1:
+            warning(f"SHB Major version {major} unsupported !")
+            raise EOFError
+
+        # Skip minor version & section length
+        self.f.seek(10, 1)
+
+        options = self.f.read(blocklen - 28)
         self._read_options(options)
+        self._read_block_tail(blocklen)
```
Comment (jcpvdm) [inline scapy/utils.py:1658]: @guedou Here I swapped the position of `self._read_block_tail(blocklen)` unintentionally. I think it's not making a big difference, but it was another unnecessary change.
By the way, should I proceed to make the changes here and your other suggestions? If yes, how should I do it? squash to a single commit (via git rebase? or git reset --soft?), or apply incremental ones?

**Comment by guedou** [comment]
Unless specifically asked or discussed during a PR review process, we squash all commits. You can simply add more commits to apply changes.

File: test/regression.uts (hunk 2 above)
Comment (guedou) [reply test/regression.uts:2456]: I think that you should keep the old `pcap` tag the way there are.

File: scapy/utils.py (hunk 3 above)
Comment (guedou) [reply scapy/utils.py:1642]: You are right this is a mistake.

**Comment by jcpvdm** [comment]
@guedou, I have done some changes as per your review suggestion. Additionally added a length check for the tsresol option when reading IDB.

File: test/regression.uts (hunk 2 above)
Comment (jcpvdm) [reply test/regression.uts]: ok, reverted the existing ones. the new pcap tag can stay under test-set or shall be changed/removed?

Comment (guedou) [reply test/regression.uts]: You can keep the new tag. Thanks.

**Review by guedou — APPROVED** [review]
Looks good to me. @gpotter2 can you do a second review?

**Review by gpotter2 — APPROVED** [review]
Thanks for the PR ! This looks good
